### PR TITLE
NETBEANS-4035: CDI Bean as Named is not listed in xhtml for selection

### PR DESCRIPTION
**Pull request by juneau001**
This PR adds support for CDI injection in Jakarta EE 9+ projects.  This will allow @Named beans to be injectable within JSF views via expression language.

---
**^Add meaningful description above**

By opening a pull request you confirm that, unless explicitly stated otherwise, the changes -

 - are all your own work, and you have the right to contribute them.
 - are contributed solely under the terms and conditions of the Apache License 2.0 (see section 5 of the license for more information).

Please make sure (eg. `git log`) that all commits have a valid name and email address for you in the Author field.

If you're a first time contributor, see the Contributing guidelines for more information.

If you're a committer, please label the PR before pressing "Create pull request" so that the right test jobs can run.

### CODE REVIEW

**Comment by neilcsmith-net** [comment]
Did you mean to put NB18 on this?  It's open against master, not delivery, and would seem a lot to land after feature freeze.

Added do-not-merge label for now pending clarification and review.

**Comment by matthiasblaesing** [comment]
@juneau001 to make review easier: Am I right if I suspect, that `jakarta.web.beans` was created by copying `web.beans` and then modifying the resulting copy (different namespace for example)?

**Comment by juneau001** [comment]
Hi @matthiasblaesing and @neilcsmith-net I did wish to try and get this fix into NB18, but I certainly can understand if it needs to be pushed to NB19 due to the timing.  You are correct @matthiasblaesing that jakarta.web.beans is a replica of web.beans with the namespaces changed.  I also made some code adjustments within other modules to determine which of the web beans modules to load on a per-project basis.

If you wish to try and get this into NB18, then I can work on opening a different pull request against the delivery branch.  Otherwise, I can re-label to NB19 if you'd rather hold off for now...it is certainly understandable.  Thanks for the feedback!

**Comment by pepness** [comment]
Hi @juneau001, I submit a PR to your branch.

File: enterprise/web.jsf/src/org/netbeans/modules/web/jsf/wizards/ManagedBeanPanelVisual.java [hunk 1]
```diff
@@ -93,8 +95,13 @@ public ManagedBeanPanelVisual(Project proj) {
             }
         }
         Object[] scopes;
-        CdiUtil cdiUtil = proj.getLookup().lookup(CdiUtil.class);
-        isCDIEnabled = cdiUtil != null && cdiUtil.isCdiEnabled();
+        if(profile != null && (profile.isAtLeast(Profile.JAKARTA_EE_9_WEB) || profile.isAtLeast(Profile.JAKARTA_EE_9_FULL))){
```
Comment (matthiasblaesing) [inline enterprise/web.jsf/src/org/netbeans/modules/web/jsf/wizards/ManagedBeanPanelVisual.java:98]: The check for the full profile should not be necessary. 
- 9 is it in both cases
- `WEB` is lower than `FULL`

```java
        if(profile != null && profile.isAtLeast(Profile.JAKARTA_EE_9_WEB)){
```

File: enterprise/web.jsf/src/org/netbeans/modules/web/jsf/wizards/PersistenceClientIterator.java [hunk 2]
```diff
@@ -632,8 +632,18 @@ private static boolean showImportStatement(String packageName, String fqn) {
     }
 
     private static boolean isCdiEnabled(Project project) {
-        CdiUtil cdiUtil = project.getLookup().lookup(CdiUtil.class);
-        return (cdiUtil == null) ? false : cdiUtil.isCdiEnabled();
+        WebModule wm = WebModule.getWebModule(project.getProjectDirectory());
```
Comment (matthiasblaesing) [inline enterprise/web.jsf/src/org/netbeans/modules/web/jsf/wizards/PersistenceClientIterator.java:635]: Can really be both `CdiUtils` in the lookup? If not, I would ditch the check here and just try to lookup both sequentially and the first that is found, is used.

File: ide/languages.hcl/nbproject/project.properties [hunk 3]
```diff
@@ -14,5 +14,5 @@
 # KIND, either express or implied.  See the License for the
 # specific language governing permissions and limitations
 # under the License.
-javac.source=1.8
+javac.source=11
```
Comment (matthiasblaesing) [inline ide/languages.hcl/nbproject/project.properties:17]: This looks strange and wrong.

File: java/java.lsp.server/nbcode/nbproject/platform.properties [hunk 4]
```diff
@@ -344,6 +343,7 @@ disabled.modules=\
     org.netbeans.modules.j2ee.sun.appsrv,\
     org.netbeans.modules.j2ee.sun.dd,\
     org.netbeans.modules.j2ee.sun.ddui,\
+    org.netbeans.modules.jakartaee.web.beans,\
```
Comment (matthiasblaesing) [inline java/java.lsp.server/nbcode/nbproject/platform.properties:346]: This looks strange (both changes). Unintentional?

File: nbbuild/cluster.properties [hunk 5]
```diff
@@ -763,6 +762,7 @@ nb.cluster.enterprise=\
         j2ee.sun.ddui,\
         j2eeapis,\
         j2eeserver,\
+        jakarta.web.beans,\
```
Comment (matthiasblaesing) [inline nbbuild/cluster.properties:765]: Why did you remove `languages.hcl`?

**Review by matthiasblaesing — COMMENTED** [review]
Thank you for the work - it is great, that you are working on this.

When building from source, I noticed, that I can't activate the java cluster or the java ee cluster in the plugin manager, so I could not have a look at the new situation.

For the naming I would suggest to try to keep the directories together, so instead of `jakarta.web.beans`, I'd suggest `web.beans.jakarta`.

I noticed that the new module is named identically to the original one, I think some kind of suffix should be added to discern the two. The same might be required for templates and similar stuff.

Please don't take this negatively, the work is appreciated.

File: ide/languages.hcl/nbproject/project.properties (hunk 3 above)
Comment (juneau001) [reply ide/languages.hcl/nbproject/project.properties:17]: I made this adjustment because my local build would not compile otherwise.  I'll take another look and revert if that works.

File: java/java.lsp.server/nbcode/nbproject/platform.properties (hunk 4 above)
Comment (juneau001) [reply java/java.lsp.server/nbcode/nbproject/platform.properties:346]: I will certainly review this adjustment.

File: nbbuild/cluster.properties (hunk 5 above)
Comment (juneau001) [reply nbbuild/cluster.properties:765]: I'm did not intentionally remove it.  Thanks for the review.  I will make the adjustments to add back.

**Comment by juneau001** [comment]
> Thank you for the work - it is great, that you are working on this.
> 
> When building from source, I noticed, that I can't activate the java cluster or the java ee cluster in the plugin manager, so I could not have a look at the new situation.
> 
> For the naming I would suggest to try to keep the directories together, so instead of `jakarta.web.beans`, I'd suggest `web.beans.jakarta`.
> 
> I noticed that the new module is named identically to the original one, I think some kind of suffix should be added to discern the two. The same might be required for templates and similar stuff.
> 
> Please don't take this negatively, the work is appreciated.

@matthiasblaesing I was thinking that it may make sense to keep all of the new jakarta packages together.  It makes sense to keep web.beans and web.beans.jakarta together for now, but thinking of the long term does it make sense to keep this new package named jakarta.web.beans in an effort to group jakarta-specific packages together?  Thanks

File: enterprise/web.jsf/src/org/netbeans/modules/web/jsf/wizards/PersistenceClientIterator.java (hunk 2 above)
Comment (juneau001) [reply enterprise/web.jsf/src/org/netbeans/modules/web/jsf/wizards/PersistenceClientIterator.java:635]: I believe we would receive a class not found exception if using Jakarta EE 9+ and returning a org.netbeans.modules.web.beans.CdiUtil.class, so I think it is a good idea to perform the check to see which version of Jakarta EE/Java EE is in use by the project.

Comment (matthiasblaesing) [reply enterprise/web.jsf/src/org/netbeans/modules/web/jsf/wizards/PersistenceClientIterator.java:635]: I don't see an issue, this should work:

```java
    private static boolean isCdiEnabled(Project project) {
        org.netbeans.modules.jakarta.web.beans.CdiUtil jakartaCdiUtil = project.getLookup().lookup(org.netbeans.modules.jakarta.web.beans.CdiUtil.class);
        if(jakartaCdiUtil != null && jakartaCdiUtil.isCdiEnabled()) {
            return true;
        }
        org.netbeans.modules.web.beans.CdiUtil javaxCdiUtil = project.getLookup().lookup(org.netbeans.modules.web.beans.CdiUtil.class);
        if(javaxCdiUtil != null && javaxCdiUtil.isCdiEnabled()) {
            return true;
        }
        return false;
    }
```

If the classes could not be loaded, it would blow up anyway as `PersistenceClientIterator` could not be loaded then. For exceptions from the `lookup` method:

https://bits.netbeans.org/17/javadoc/org-openide-util-lookup/org/openide/util/Lookup.html#lookup-java.lang.Class-

> **Returns:**
> an object implementing the given class or null if no such implementation is found

So if the corresponding CdiUtil is present, then it is used else it is null and not checked.

**Comment by matthiasblaesing** [comment]
> @matthiasblaesing I was thinking that it may make sense to keep all of the new jakarta packages together. It makes sense to keep web.beans and web.beans.jakarta together for now, but thinking of the long term does it make sense to keep this new package named jakarta.web.beans in an effort to group jakarta-specific packages together?

Understood. My reasoning would be that javamail and jakarta-mail or web.beans and jakarta.web.beans have more in common with each other, than jakarta mail and jakarta.web.beans. On the other hand, this is a side discussion and was just my first impression. If you follow my reasoning great, if not, also ok, I'd like to see jakarta support in NetBeans and that is the important point ;-)

File: enterprise/web.jsf/src/org/netbeans/modules/web/jsf/wizards/PersistenceClientIterator.java (hunk 2 above)
Comment (juneau001) [reply enterprise/web.jsf/src/org/netbeans/modules/web/jsf/wizards/PersistenceClientIterator.java:635]: I see what you are explaining now and that will work well.  I appreciate the feedback.

**Comment by mbien** [comment]
as @neilcsmith-net mentioned before. It needs clarification if this PR is meant to get into NB 18 or not.
- if not -> change milestone to NB 19 please
- if it is intended to get into NB 18, it has to be rebased onto the `delivery` branch and the target has to be changed to `delivery` too

I am fairly neutral on this, but since jakarta support is a bit limited atm, I would be ok in taking the risk to integrate a larger PR like this one here even during the RC phase. Usually we would try to integrate only bug fixes during the RC phase.

**Comment by juneau001** [comment]
I believe I have addressed all of the comments with the exception of renaming from jakarta.web.beans to web.beans.jakarta.  In an effort to position the jakarta specific packages together, I am going to leave it as jakarta.web.beans.  Also added some commits from @pepness into this PR.  Thanks for the contributions.

**Comment by matthiasblaesing** [comment]
I think it would be better to get this into NB19 early and give it bit of time to settle. This is IMHO still work in progress and not ready for merging.

**Review by matthiasblaesing — COMMENTED** [review]
@juneau001 thank you for the updates. Much looks good to me now, but I think you overlooked some comments. I marked all comments as resolved, that I think were either addressed or where you explained the decisions.

While looking through the changes I noticed two more things, which I left inline.

Running a build from source with `ant tryme` gives the the following messages.log:
[messages.log.zip](https://github.com/apache/netbeans/files/11334964/messages.log.zip). What I notice:

Warnings from ergonomics:

```
WARNING [org.netbeans.core.projects.cache]: layer jar:file:/home/matthias/src/netbeans/nbbuild/netbeans/ergonomics/modules/org-netbeans-modules-ide-ergonomics.jar!/org/netbeans/modules/ide/ergonomics/enterprise/layer.xml contains duplicate attributes SystemFileSystem.localizingBundle for Templates/CDI/beans.xml
```

I suspect, that ergonomics pulls all string ids together and this results in these warnings, as the strings are now in two modules.

```
WARNING [org.openide.filesystems.Ordering]: Found same position 799 for both Services/MIMEResolver/org-netbeans-modules-jakarta-web-beans-BeansDataObject-Namespace.xml and Services/MIMEResolver/org-netbeans-modules-web-beans-BeansDataObject-Namespace.xml
```

The position of the MimeResolver should be adjusted for the new module, so that the to entries don't occupy the same position.

```
Caused: java.io.FileNotFoundException: jar:file:/home/matthias/src/netbeans/nbbuild/netbeans/enterprise/modules/org-netbeans-modules-jakarta-web-beans.jar!/org/netbeans/modules/jakarta/web/beans/resources/org-netbeans-modules-jakarta-web-beans-annotations-intercepted.xml
	at org.netbeans.JarClassLoader$NbJarURLConnection.connect(JarClassLoader.java:1187)
	at org.netbeans.JarClassLoader$NbJarURLConnection.getInputStream(JarClassLoader.java:1223)
	at org.netbeans.core.startup.layers.BinaryFS$BFSFile.getInputStream(BinaryFS.java:863)
Caused: java.io.FileNotFoundException: Cannot find 'jar:file:/home/matthias/src/netbeans/nbbuild/netbeans/enterprise/modules/org-netbeans-modules-jakarta-web-beans.jar!/org/netbeans/modules/jakarta/web/beans/resources/org-netbeans-modules-jakarta-web-beans-annotations-intercepted.xml'
	at org.netbeans.core.startup.layers.BinaryFS$BFSFile.getInputStream(BinaryFS.java:866)
	at org.openide.filesystems.MIMESupport$CachedFileObject.getInputStream(MIMESupport.java:377)
[catch] at org.netbeans.modules.openide.filesystems.declmime.DefaultParser.parse(DefaultParser.java:115)
	at org.netbeans.modules.openide.filesystems.declmime.XMLMIMEComponent$SniffingParser.sniff(XMLMIMEComponent.java:269)
	at org.netbeans.modules.openide.filesystems.declmime.XMLMIMEComponent.acceptFileObject(XMLMIMEComponent.java:81)
	at org.netbeans.modu
```

this looks as if files are missing (this is repeated for several files).

**Comment by matthiasblaesing** [comment]
@juneau001 I gave the latest version of this PR a spin and noticed warnings and also saw the commit validation test failing here. I had a quick look and indeed the layer file has duplicate entries. This patch fixes this:

[warnings.patch.zip](https://github.com/apache/netbeans/files/11445353/warnings.patch.zip)

If I'm not mistaken though, this is still not quite correct, as I was in a JavaEE project and asked NetBeans to add a JakarataEE beans.xml, but got a JavaEE one. While that is correct for the project, it is strange to see two template entries then. Anyway, maybe the patch can help to push this further.

**Comment by juneau001** [comment]
> @juneau001 I gave the latest version of this PR a spin and noticed warnings and also saw the commit validation test failing here. I had a quick look and indeed the layer file has duplicate entries. This patch fixes this:
> 
> [warnings.patch.zip](https://github.com/apache/netbeans/files/11445353/warnings.patch.zip)
> 
> If I'm not mistaken though, this is still not quite correct, as I was in a JavaEE project and asked NetBeans to add a JakarataEE beans.xml, but got a JavaEE one. While that is correct for the project, it is strange to see two template entries then. Anyway, maybe the patch can help to push this further.

Thanks for the patch...I will apply it.  Actually, the fix in this PR does not repair the incorrect beans.xml namespaces.  The code changes in this PR repair the code completion issues only.  I am working on another PR that will repair XHTML namespaces for Jakarta Faces 4+.  I can also look into repairing the beans.xml namespaces.

**Comment by matthiasblaesing** [comment]
@pepness @juneau001 I think it would be good to squash this into a single commit. We had the situation in the past where it was necessary to revert a change and having a disconnected set of commits with intermittent merges makes that hard.

Locally I did this:

- `git fetch github`
- `git checkout -b pr-5866 github/pr/5866`
- `git rebase master`
- added my own changes
- `git rebase -i HEAD~9`

Marked all commits as "squash", saved and updated the commit message to read

```
    NETBEANS-4035: CDI Bean as Named is not listed in xhtml for selection
    
    Closes: #4035
    Co-authored-by: José Contreras <pepness@apache.org>
    Co-authored-by: Matthias Bläsing <mblaesing@doppel-helix.eu>
```

and retained @juneau001 as author of the commit.

The result can be observed here: https://github.com/matthiasblaesing/netbeans/commit/697d0320eaffbd054813a5d0c369bf5ebae47703

What do you think about this?

**Comment by juneau001** [comment]
> @pepness @juneau001 I think it would be good to squash this into a single commit. We had the situation in the past where it was necessary to revert a change and having a disconnected set of commits with intermittent merges makes that hard.
> 
> Locally I did this:
> 
> * `git fetch github`
> * `git checkout -b pr-5866 github/pr/5866`
> * `git rebase master`
> * added my own changes
> * `git rebase -i HEAD~9`
> 
> Marked all commits as "squash", saved and updated the commit message to read
> 
> ```
>     NETBEANS-4035: CDI Bean as Named is not listed in xhtml for selection
>     
>     Closes: #4035
>     Co-authored-by: José Contreras <pepness@apache.org>
>     Co-authored-by: Matthias Bläsing <mblaesing@doppel-helix.eu>
> ```
> 
> and retained @juneau001 as author of the commit.
> 
> The result can be observed here: [matthiasblaesing@697d032](https://github.com/matthiasblaesing/netbeans/commit/697d0320eaffbd054813a5d0c369bf5ebae47703)
> 
> What do you think about this?

Hi @matthiasblaesing have you already performed these steps?  Is there anything you need me to do?  Everything looks good and I agree with your result.  Thanks for your time and assistance.

**Comment by matthiasblaesing** [comment]
> Hi @matthiasblaesing have you already performed these steps? Is there anything you need me to do? Everything looks good and I agree with your result. Thanks for your time and assistance.

Yes, the referenced branch was created by this. You allowed edits for this PR, so I could force push the updated branch into the branch that is backing this PR. We would then see whether unittests all run green and if so merge.

Please indicate whether I should do that.

**Comment by juneau001** [comment]
> 

Hi @matthiasblaesing.  Thanks for the response.  Yes, please feel free to move forward.  I appreciate your time and assistance!

**Comment by matthiasblaesing** [comment]
Tests are green, will merge tomorrow unless someone objects.

**Comment by IbrahimMSoliman** [comment]
Thank you so much for fixing this issue

**Comment by matthiasblaesing** [comment]
@juneau001 @pepness thank you both for working on this!

**Comment by juneau001** [comment]
> @juneau001 @pepness thank you both for working on this!

Thank you both @pepness @matthiasblaesing  for your time and assistance!

**Comment by neilcsmith-net** [comment]
Please can you do a double check on module versions and sig files in this, thanks!  There were a few conflicts while preparing an updated version of #5856   It looks like at least one module version has gone backwards (`web.beans`).  It should rarely be necessary to touch sig files or module versions between releases.
